src/agent/block_store_services/block_store_fs.js

```diff
@@ -58,171 +60,181 @@ class BlockStoreFs extends BlockStoreBase {
 
-        return Promise.resolve()
-            .then(() => fs.promises.stat(block_path))
-            .catch(ignore_not_found)
-            .then(stat => {
-                overwrite_stat = stat;
-                dbg.log1('_write_block', block_path, data.length, overwrite_stat);
-                // create/replace the block on fs
-                return Promise.all([
-                    fs.promises.writeFile(block_path, data),
-                    fs.promises.writeFile(meta_path, block_md_data),
-                ]);
-            })
-            .catch(err => {
-                if (err.code === 'ENOENT') {
-                    dbg.error('got error when writing', block_md, '. checking if root_path exists', err.message);
-                    this._test_root_path_exists();
-                }
-                throw err;
-            })
-            .then(() => {
-                if (overwrite_stat) {
-                    return fs.promises.stat(meta_path).catch(ignore_not_found)
-                        .then(md_stat => {
-                            md_overwrite_stat = md_stat;
-                        });
-                }
-            })
-            .then(() => {
-                let overwrite_size = 0;
-                let overwrite_count = 0;
-                if (overwrite_stat) {
-                    overwrite_size = overwrite_stat.size + (md_overwrite_stat ?
-                        md_overwrite_stat.size : 0);
-                    overwrite_count = 1;
+        if (!is_test_block) {
+
+            // set xattr to trigger migration of file to underlying tier
+            if (config.BLOCK_STORE_FS_TIER2_ENABLED) {
+                xattr[config.BLOCK_STORE_FS_XATTR_TRIGGER_MIGRATE] = 'now';
+                xattr_need_fsync = true;
+            }
+
+            // check for existing file and get its size
+            const overwrite_stat = await nb_native().fs.stat(fs_context, block_path).catch(ignore_not_found);
+            if (overwrite_stat) {
+                usage.size -= overwrite_stat.size;
+                usage.count -= 1;
+
+                // also make sure we do not leave old .meta files on overwrite
+                const meta_path = this._get_block_meta_path(block_md.id);
+                const overwrite_meta_stat = await nb_native().fs.stat(fs_context, meta_path).catch(ignore_not_found);
+                if (overwrite_meta_stat) {
+                    usage.size -= overwrite_meta_stat.size;
+                    await nb_native().fs.unlink(fs_context, meta_path).catch(ignore_not_found);
                 }
-                const size = (block_md.is_preallocated ? 0 : data.length) + block_md_data.length - overwrite_size;
-                const count = (block_md.is_preallocated ? 0 : 1) - overwrite_count;
-                if (size || count) this._update_usage({ size, count });
+            }
+        }
+
+        try {
+            await nb_native().fs.writeFile(fs_context, block_path, data, {
```

Comment:
I am just trying to understand here. What happens if the unlink succeeds but this fails and `_test_root_path_exists` throws as well?

Reply:
The usage is not going to update in that case, which is indeed undesired, but since usage calculation is not atomic anyhow, any failure can cause this.

I explained to @romayalon's question in the [comment above](https://github.com/noobaa/noobaa-core/pull/7283/files#r1185765251) that the overwrite block case is not an actual flow we invoke on purpose.

The reason we decided that write_block should handle overwrites was that we didn't want to fail user uploads on one hand, and we didn't want the usage to start getting skewed because of retries so we had to check if this is an overwrite first and "reduce" that replaced size.

I guess that simply failing the request if the same block is written with different data could just be better.

Reply:
For now I will keep the existing behavior and we can revise as needed in the future.